### PR TITLE
SNOW-1956120: handle none data and improve upload logic

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-NNNNNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/data_source_utils.py [hunk 1]
```diff
@@ -283,7 +286,10 @@ def infer_data_source_schema(
 
 def data_source_data_to_pandas_df(data: List[Any], schema: StructType) -> pd.DataFrame:
     columns = [col.name for col in schema.fields]
-    df = pd.DataFrame.from_records(data, columns=columns)
+    if not data:
+        return pd.DataFrame(columns=columns)
```
Comment (sfc-gh-yuwang) [inline src/snowflake/snowpark/_internal/data_source_utils.py:290]: is there a reason we did this special treatment? I tried `pd.DataFrame([],columns=["a", "b"])`, and it does not produce any error

File: src/snowflake/snowpark/_internal/data_source_utils.py [hunk 2]
```diff
@@ -283,7 +286,10 @@ def infer_data_source_schema(
...
+    # this way handles both list of object and list of tuples and avoid implict pandas type conversion
+    df = pd.DataFrame([list(row) for row in data], columns=columns, dtype=object)
```
Comment (sfc-gh-yuwang) [inline src/snowflake/snowpark/_internal/data_source_utils.py:292]: I think we also need to test if there is any perf degrade after we do this

File: src/snowflake/snowpark/dataframe_reader.py [hunk 3]
```diff
@@ -1511,15 +1513,19 @@ def convert_to_parquet(fetched_data, fetch_idx):
...
             cursor = cursor.execute(query)
             fetch_idx = 0
             while True:
                 rows = cursor.fetchmany(fetch_size)
                 if not rows:
                     break
-                parquet_file_queue.put(convert_to_parquet(rows, fetch_idx))
+                parquet_file_path = convert_to_parquet(rows, fetch_idx)
+                if parquet_file_path:
+                    parquet_file_queue.put(parquet_file_path)
```
Comment (sfc-gh-yuwang) [inline src/snowflake/snowpark/dataframe_reader.py:1528]: we should use logger or other form to tell user that their data source is empty, otherwise user will only see they run dbapi and there is no data fetched

File: tests/integ/test_data_source_api.py [hunk 4]
```diff
@@ -169,32 +169,34 @@ def test_dbapi_retry(session):
     IS_WINDOWS,
     reason="sqlite3 file can not be shared across processes on windows",
 )
-def test_parallel(session):
+@pytest.mark.parametrize("upper_bound, expected_upload_cnt", [(5, 3), (100, 1)])
```
Comment (sfc-gh-yuwang) [inline tests/integ/test_data_source_api.py:172]: we should also include a test of behavior when the target table is completely empty/ all data are null

File: src/snowflake/snowpark/_internal/data_source_utils.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/_internal/data_source_utils.py:307]: no special reason just my own code preference, I can consolidate the two

File: src/snowflake/snowpark/_internal/data_source_utils.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/_internal/data_source_utils.py:309]: good call out, doing it now

File: src/snowflake/snowpark/dataframe_reader.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/dataframe_reader.py:1548]: we can do debug log

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/dataframe_reader.py:1551]: added debug log in convert_to_parquet

File: tests/integ/test_data_source_api.py (hunk 4 above)
Comment (sfc-gh-aling) [reply tests/integ/test_data_source_api.py:173]: test_empty_table

**Review by sfc-gh-yuwang — APPROVED** [review]

File: src/snowflake/snowpark/_internal/data_source_utils.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/_internal/data_source_utils.py:307]: test with no perf degrade

File: src/snowflake/snowpark/dataframe_reader.py [hunk 5]
```diff
@@ -1513,6 +1513,11 @@ def _task_fetch_from_data_source(
     ):
         def convert_to_parquet(fetched_data, fetch_idx):
             df = data_source_data_to_pandas_df(fetched_data, schema)
+            if df.empty:
+                logger.debug(
+                    f"The DataFrame is empty, no parquet file is generated for partition {partition_idx} fetch {fetch_idx}."
+                )
+                return ""
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/dataframe_reader.py:1520]:
```suggestion
                return None
```
